### PR TITLE
mem: Don't use posix_memalign() and friends with custom wrapper.

### PR DESCRIPTION
**Pull request by sebastianas**
If the application provides custom memory allocations functions via CRYPTO_set_mem_functions() then those should be used instead something else like posix_memalign(). The applications might verify alloc and free calls and pointers from posix_memalign() were never returned by the implementations.

At least stunnel4 complains here.

Use posix_memalign() or if aligned_alloc() only if the application did not provide a custom malloc() implementation. In case of a custom implementation use CRYPTO_malloc() and align the memory accordingly.

Fixes: #25678

### CODE REVIEW

File: crypto/mem.c [hunk 1]
```diff
@@ -238,15 +238,21 @@ void *CRYPTO_aligned_alloc(size_t num, size_t alignment, void **freeptr,
...
 
-#if defined (_BSD_SOURCE) || (defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 200112L)
-    if (posix_memalign(&ret, alignment, num))
-        return NULL;
-    *freeptr = ret;
-    return ret;
+    /*
+     * Allow non-malloc() allocations as long as no malloc_impl is provided.
+     */
+    if (!malloc_impl) {
```
Comment (slontis) [inline crypto/mem.c:244]:
```suggestion
    if (!malloc_impl != NULL) {
```

File: crypto/mem.c [hunk 2]
```diff
@@ -238,15 +238,21 @@ void *CRYPTO_aligned_alloc(size_t num, size_t alignment, void **freeptr,
     return ret;
 #endif
 
-#if defined (_BSD_SOURCE) || (defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 200112L)
-    if (posix_memalign(&ret, alignment, num))
-        return NULL;
-    *freeptr = ret;
-    return ret;
+    /*
+     * Allow non-malloc() allocations as long as no malloc_impl is provided.
```
Comment (slontis) [inline crypto/mem.c:242]:
```suggestion
     /* Allow non-malloc() allocations as long as no malloc_impl is provided. */
```

**Review by Sashan — APPROVED** [review]
I agree with those changes. Thanks for a bug report and fix.

**Review by mattcaswell — APPROVED** [review]

**Review by t8m — APPROVED** [review]

File: crypto/mem.c (hunk 2 above)
Comment (sebastianas) [reply crypto/mem.c]: Thanks, updated as suggested.

**Review by shahsb — APPROVED** [review]

**Review by slontis — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master and 3.4 branches. Thank you for your contribution.

**Pull request closed without merging**